packages/filebrowser-extension/src/index.ts

```diff
@@ -351,104 +419,101 @@ function createMenu(app: JupyterLab, creatorCmds: string[]): Menu {
 
 
 /**
- * Create a context menu for the file browser listing.
+ * Populate the context menu for the file browser listing.
  *
  * #### Notes
- * This function generates temporary commands with an incremented name. These
- * commands are disposed when the menu itself is disposed.
+ * The `"Open With"` menu is dynamically added/removed as the selection on
+ * the file browser changes.
  */
-function createContextMenu(fbWidget: FileBrowser, openWith: Menu):  Menu {
-  let { commands } = fbWidget;
-  let menu = new Menu({ commands });
-  let prefix = `file-browser-${++Private.id}`;
-  let disposables = new DisposableSet();
-  let command: string;
-
-  // Remove all the commands associated with this menu upon disposal.
-  menu.disposed.connect(() => { disposables.dispose(); });
-
-  command = `${prefix}:open`;
-  disposables.add(commands.addCommand(command, {
-    execute: () => { fbWidget.open(); },
-    icon: 'jp-MaterialIcon jp-OpenFolderIcon',
-    label: 'Open',
-    mnemonic: 0
-  }));
-  menu.addItem({ command });
+function populateContextMenu(fbWidget: FileBrowser, menu: ContextMenu, commands: CommandRegistry) {
+  let selector = '.jp-DirListing-content';
+  menu.addItem({ command: CommandIDs.open, selector, rank: 1 });
+
+  let openWith = new Menu({ commands });
+  openWith.title.label = 'Open With...';
+  let submenuSelector = `${selector}.jp-mod-open-with`;
+  menu.addItem({
+    type: 'submenu', submenu: openWith, selector: submenuSelector, rank: 1
+  });
 
-  if (openWith) {
-    menu.addItem({ type: 'submenu', submenu: openWith });
-  }
+  fbWidget.selectionChanged.connect(() => {
+    openWith.clearItems();
+    let anyEnabled = false;
+    each(fbWidget.registry.widgetFactories(), widgetFactory => {
+      let factory = widgetFactory.name;
+      if (commands.isVisible(CommandIDs.open, { factory })) {
+        openWith.addItem({ command: CommandIDs.open, args: { factory } });
+      }
+      if (commands.isEnabled(CommandIDs.open, { factory })) {
+        anyEnabled = true;
+      }
+    });
```

Comment:
This is brittle logic. It doesn't account for those commands being hidden or disabled after a selection change. We may need to talk about this some more.